include/Gaffer/Private/IECorePreview/TaskMutex.h

```diff
@@ -0,0 +1,357 @@
+					return m_lock.upgrade_to_writer();
+				}
+
+				/// Calls `f` in a way that allows threads waiting for the lock to perform
+				/// TBB tasks on its behalf. Should only be called by the holder of a write lock.
+				template<typename F>
+				void execute( F &&f )
+				{
+					assert( m_mutex && m_writer && !m_recursive );
+
+					ExecutionStateMutex::scoped_lock executionStateLock( m_mutex->m_executionStateMutex );
+					assert( !m_mutex->m_executionState );
+					m_mutex->m_executionState = std::make_shared<ExecutionState>();
+					executionStateLock.release();
+
+					m_mutex->m_executionState->arena.execute(
+						[this, &f] {
+							// Note : We deliberately call `run()` and `wait()` separately
+							// instead of calling `run_and_wait()`. The latter is buggy until
+							// TBB 2018 Update 3, causing calls to `wait()` on other threads to
+							// return immediately rather than do the work we want.
+							m_mutex->m_executionState->taskGroup.run( f );
+							m_mutex->m_executionState->taskGroup.wait();
+						}
+					);
+
+					executionStateLock.acquire( m_mutex->m_executionStateMutex );
+					m_mutex->m_executionState = nullptr;
+				}
+
+				/// Acquires mutex or returns false. Never does TBB tasks.
+				bool tryAcquire( TaskMutex &mutex, bool write = true )
+				{
+					return acquireOr( mutex, write, []( bool workAvailable ){ return false; } );
+				}
+
+				/// Releases the lock. This will be done automatically
+				/// by ~ScopedLock, but may be called explicitly to release
+				/// the lock early.
+				void release()
+				{
+					assert( m_mutex );
+					if( !m_recursive )
+					{
+						m_lock.release();
+					}
+					m_mutex = nullptr;
+				}
+
+				/// Advanced API
+				/// ============
+				///
+				/// These methods provide advanced usage required by complex requirements
+				/// in Gaffer's LRUCache. They should not be considered part of the canonical
+				/// API.
+
+				/// Returns true if `acquire()` obtained a recursive lock rather
+				/// than a unique lock. Recursive locks are available to any thread
+				/// performing work on behalf of `execute()`.
+				bool recursive() const
+				{
+					return m_recursive;
+				}
+
+				/// Tries to acquire the mutex, returning true on success. On failure,
+				/// calls `workNotifier( bool workAvailable )`. If work is available and
+				/// `workNotifier` returns true, then this thread will perform TBB tasks
+				/// spawned by `execute()` until the work is complete. Returns false on
+				/// failure regardless of whether or not work is done.
+				template<typename WorkNotifier>
+				bool acquireOr( TaskMutex &mutex, bool write, WorkNotifier &&workNotifier )
+				{
+					assert( !m_mutex );
+					if( m_lock.try_acquire( mutex.m_mutex, write ) )
+					{
+						// Success!
+						m_mutex = &mutex;
+						m_recursive = false;
+						m_writer = write;
+						return true;
+					}
+
+					ExecutionStateMutex::scoped_lock executionStateLock( mutex.m_executionStateMutex );
```

Comment:
this block might do with a comment explaining what we're aiming for. the comment on `acquireOr` doesn't quite explain the 2 attempts at acquiring a mutex, and we haven't yet learned what an `ExecutionStateMutex` is

Reply:
Added further comments in 374fca8, that hopefully go some way to clarifying this.